File: src/main/kotlin/java/_02_data_calculations_actions/couponDog/Main.java
package java._02_data_calculations_actions.couponDog;

import java._02_data_calculations_actions.couponDog.actions.DatabaseService;
import java._02_data_calculations_actions.couponDog.actions.EmailService;
import java._02_data_calculations_actions.couponDog.data.Coupon;
import java._02_data_calculations_actions.couponDog.data.CouponRank;
import java._02_data_calculations_actions.couponDog.data.Email;
import java._02_data_calculations_actions.couponDog.data.Subscriber;

import java.util.ArrayList;
import java.util.List;

public class Main {

    public static void main(String[] args) {
//        List<String> couponList = List.of("10% off", "20% off", "30% off", "40% off", "50% off");
//
//        var best = couponList.stream().filter(coupon -> coupon.contains("50%")).toList();
//        for ( String coupon : best ) {
//            System.out.println(coupon);
//        }

        sendNewsletter();
    }

    public static void sendNewsletter() {
        var coupons = DatabaseService.fetchCoupons();
        var goodCoupons = selectCouponsByRank(coupons, CouponRank.GOOD);
        var bestCoupons = selectCouponsByRank(coupons, CouponRank.BEST);
        var subscribers = DatabaseService.fetchSubscribers();
        var emails = couponEmailsForSubscribers(subscribers, goodCoupons, bestCoupons);
        EmailService.sendEmails(emails);
    }

    public static List<Email> couponEmailsForSubscribers(List<Subscriber> subscribers, List<Coupon> goodCoupons, List<Coupon> bestCoupons) {
        var emails = new ArrayList<Email>();
        for (Subscriber subscriber : subscribers) {
            emails.add(couponEmailForSubscriber(subscriber, goodCoupons, bestCoupons));
        }
        return emails;
    }

    public static Email couponEmailForSubscriber(Subscriber subscriber, List<Coupon> goodCoupons, List<Coupon> bestCoupons)
    {
        return switch (subscriberCouponRank(subscriber)) {
            case GOOD -> makeEmail(subscriber, goodCoupons);
            case BEST -> makeEmail(subscriber, bestCoupons);
            default -> throw new RuntimeException("We should never get here!");
        };
    }

    public static Email makeEmail(Subscriber subscriber, List<Coupon> coupons) {
        var couponStrings = toStringList(coupons);
        return new Email("newsletter@coupondog.ca",
                subscriber.email(),
                "Coupons for you!",
                "Hi there, " + subscriber.email() + "! Here are some coupons for you: \n" + String.join("\n", couponStrings) + "\n");
    }

    public static List<String> toStringList(List<Coupon> coupons) {
//        var couponStrings = new ArrayList<String>();
//        for (Coupon coupon : coupons) {
//            couponStrings.add(coupon.code());
//        }
//        return couponStrings;

        // Or, using streams:
        return coupons.stream().map(Coupon::code).toList();
    }

    public static CouponRank subscriberCouponRank(Subscriber subscriber) {
        return subscriber.referralCount() >= 20 ? CouponRank.BEST : CouponRank.GOOD;
    }

    public static List<Coupon> selectCouponsByRank(List<Coupon> coupons, CouponRank rank) {
//        return coupons.stream()
//                .filter(coupon -> coupon.rank().equals(rank))
//                .toList();

        // Using a loop:
        var selectedCoupons = new ArrayList<Coupon>();
        for (Coupon coupon : coupons) {
            if (coupon.rank().equals(rank)) {
                selectedCoupons.add(coupon);
            }
        }
        return selectedCoupons;
    }
}

File: src/main/kotlin/_02_data_calculations_actions/couponDog/Main.kt
package _02_data_calculations_actions.couponDog

/**
 *  NOTE: This is not a real-world example. It's a simplified example to demonstrate the concepts.
 *
 * It is also not necessary to organize your code into DATA, ACTIONS, and CALCULATIONS sections, but for the
 * sake of learning to identify these parts of the code, we've done so here.
 */

/*******************
 * DATA
 */

// Let's use enum and data classes to represent the data we're working with.
enum class CouponRank { GOOD, BAD, BEST }
enum class SubscriberRank { GOOD, BEST }
data class Coupon(val code: String, val rank: CouponRank)
data class Email(val from: String, val to: String, val subject: String, val body: String)
data class Subscriber(val email: String, val referralCount: Int)

/******************
 * CALCULATIONS
 */

fun selectCouponsByRank(coupons: List<Coupon>, rank: CouponRank) =
    coupons.filter { it.rank == rank }

fun couponEmailsForSubscribers(subscribers: List<Subscriber>, goodCoupons: List<Coupon>, bestCoupons: List<Coupon>) =
    subscribers.map { couponEmailForSubscriber(it, goodCoupons, bestCoupons) }

fun couponEmailForSubscriber(subscriber: Subscriber, goodCoupons: List<Coupon>, bestCoupons: List<Coupon>) =
    when (subscriberRank(subscriber)) {
        SubscriberRank.GOOD -> makeEmail(subscriber, goodCoupons)
        SubscriberRank.BEST -> makeEmail(subscriber, bestCoupons)
    }

fun subscriberRank(subscriber: Subscriber) =
    if (subscriber.referralCount >= 20) SubscriberRank.BEST else SubscriberRank.GOOD

fun makeEmail(subscriber: Subscriber, coupons: List<Coupon>) =
    Email(
        from = "newsletter@coupondog.com",
        to = subscriber.email,
        subject = "Check out these coupons!",
        body = "Here are your coupons: ${coupons.joinToString { it.code }}")

/********************
 * ACTIONS
 */

fun main() {
    sendNewsletter()  // This is an action, so main is an action
}

fun sendNewsletter() {
    val coupons = fetchCoupons()            // This is an action, so sendNewsletter is an action
    val subscribers = fetchSubscribers()    // This is an action, so sendNewsletter is an action

    val goodCoupons = selectCouponsByRank(coupons, CouponRank.GOOD);
    val bestCoupons = selectCouponsByRank(coupons, CouponRank.BEST);
    val emails = couponEmailsForSubscribers(subscribers, goodCoupons, bestCoupons);

    sendEmails(emails)                      // This is an action, so sendNewsletter is an action
}

// This function simulates retrieving Coupons from a database
fun fetchCoupons(): List<Coupon> {
    return listOf(
        Coupon("ABC", CouponRank.GOOD),
        Coupon("DEF", CouponRank.BAD),
        Coupon("GHI", CouponRank.BEST),
        Coupon("JKL", CouponRank.GOOD),
        Coupon("MNO", CouponRank.BAD)
    )
}

// This function simulates retrieving a Subscriber list from a database
fun fetchSubscribers(): List<Subscriber> {
    return listOf(
        Subscriber("alice@email.com", 0),
        Subscriber("bob@email.com", 23),
        Subscriber("charlie@email.com", 7),
        Subscriber("dan@email.com", 3),
        Subscriber("elise@email.com", 0),
        Subscriber("freda@email.com", 20),
        Subscriber("gertrude@email.com", 2)
    )
}

// This function simulates sending an email

fun sendEmails(emails: List<Email>) {
    for (email in emails) {
        println("Sending: ")
        println(email)
    }
}